runtime/bundles/org.eclipse.e4.core.di/src/org/eclipse/e4/core/internal/di/InjectorImpl.java

```diff
@@ -668,20 +666,14 @@ private void processClass(Object userObject, PrimaryObjectSupplier objectSupplie
 
 	private boolean hasInjectedStatic(Class<?> objectsClass) {
 		synchronized (injectedClasses) {
-			for (WeakReference<Class<?>> ref : injectedClasses) {
-				Class<?> injectedClass = ref.get();
-				if (injectedClass == null)
-					continue;
-				if (injectedClass == objectsClass) // use pointer comparison
-					return true;
-			}
-			return false;
+			int objectId = System.identityHashCode(objectsClass);
+			return injectedClasses.containsKey(objectId);
 		}
 	}
 
 	private void rememberInjectedStatic(Class<?> objectsClass) {
 		synchronized (injectedClasses) {
-			injectedClasses.add(new WeakReference<>(objectsClass));
+			injectedClasses.put(System.identityHashCode(objectsClass), new WeakReference<>(objectsClass));
```

Comment:
Is there an assumption here that there will be no collisions on this hash code?

Reply:
I thought to while implementing, but I dug a bit into it and it has collisions. 
What I found is ObjectUtils.identityToString, but that produces the same output as the default toString()
Which is ClassName@IdentitiyHashCode.

Do you have an idea how we could solve this without collisions and without always iterating the whole list?

Reply:
@merks I looked at the disposal code again. Do you agree that the order of the objects seems to be irrelevant?

My first thought was that they are disposed inverse of creation, but they are disposed in the same order, so the order should be irrelevant?

Then we could use a WeakHashMap where we use the object as key which is held in a weak reference internally. And the WeakReference as value.

Reply:
Naively I though wherever a List was being used to use instead a HashSet or LinkedHashSet if order is somehow relevant, but looking more closely now, that doesn't work because it's comparing `object == ref.get()`.

I don't see any immediately obvious reason why a weak hash map couldn't be used, but I've never looked closely at this code before...

Reply:
If object equally is relevant an `IdentityHashmap` should be used and as the code is syncronized anyways...

Reply:
@laeubi The problem is the need for weak references...

Reply:
I must confess that I don't understand the optimization goal here then, maybe it could be explained a bit more in detail. Maybe it would be enough to extend `WeakReference` and override equals/hashcode and then simply use the Set.contains, this hsould have equal performance to a Map.

Beside that System.identityHashCode() might have collisions, but this is not important as long as it is equally distributed, but you then needs to maintain a map of objects as the value and can then iterate over a much smaller set (effectively reimplementing what the first suggestion would archive with a standard set);

Reply:
@laeubi the main optimization goal is the rememberInjectedObject function.
In the original implementation it iterated over the list of all WeakReferences, resolved those and compared the object, and for a list with 400.000 elements that takes a while, basically O(n).

What I am trying is to replace this list with something hash based where a simple and fast contains check can be done.

The complicated thing is the object should be kept in a weak reference, while the contains check has to be done on the actual object.

My initial idea was to keep the hash of the object as a key to the reference, which turned out to be an issue with collusions.

That's why we are discussing which alternative we could use to store a weak reference on the object.

Reply:
As said you have two options:

1. Store `IdentityWeakReference` in a `Set<IdentityWeakReference>` where `IdentityWeakReference` is implemented as below
2. Make  `injectedClasses` a `Map<Long, List<WeakReference>>` to account for hashcode collisions

```
public class IdentityWeakReference extends WeakReference<Object> {

    public IdentityWeakReference(Object referent) {
        super(referent);
    }
    
    @Override
    public boolean equals(Object obj) {
        if (obj instanceof IdentityWeakReference other) {
            return Objects.equal(other.get(), get());
        }
        return false;
    }
    
    @Override
    public int hashCode() {
        return System.identityHashCode(get());
    }

}
```

Reply:
By the way, the current code seem to suffer from leaking `WeakReference` if the are garbage collected they could never be removed so it seems a bit misusing `WeakReference` anyways.

Reply:
Thanks for the example. I don't know where I got the idea from that there is a weakhashset, but with our first example we would build something similar. 

Should I implement this as a nested class or should this be in some reusable location where other data types are provided?

Reply:
@laeubi I updated my PR with our proposal for the `IdentityWeakReference` to keep the order of the elements, just to keep the same behavior as before, I used a LinkedHashSet.